### PR TITLE
feat: slot optimization

### PR DESCRIPTION
**Pull request by parasg1999**
The current implementation will cut off an entire slot if I'm busy for just 5 minutes at the starting of a slot.
For example, if I'm busy till 5:15pm and slot starts at 5:10pm for a 45 minute booking, then next slot should be 5:15pm instead of 5:55pm.

This PR attempts to fix that. Please give suggestions if changes are required.

### CODE REVIEW

**Comment by baileypumfleet** [comment]
No idea, why, but this seems to make Safari freeze when you select a date:

https://user-images.githubusercontent.com/25907159/117964025-06870b80-b319-11eb-958c-558c1dde336c.mov

**Comment by parasg1999** [comment]
@baileypumfleet I really don't know why it would happen, but I don't have access to the Safari browser.

**Comment by baileypumfleet** [comment]
Yeah, just tested on Chrome 90 on a completely fresh macOS machine and it freezes on Chrome too
Screenshot 2021-05-12 at 13 31 37

**Comment by parasg1999** [comment]
```js
SlotStartTime: d {$L: "en", $offset: 330, $d: Thu May 13 2021 23:59:00 GMT+0530 (India Standard Time), $x: {…}, $y: 2021, …} 

SlotEndTime: d {$L: "en", $offset: 330, $d: Fri May 14 2021 00:44:00 GMT+0530 (India Standard Time), $x: {…}, $y: 2021, …} 

BusyStartTime: d {$L: "en", $d: Thu May 13 2021 23:30:00 GMT+0530 (India Standard Time), $x: {…}, $y: 2021, $M: 4, …} 

BusyEndTime: d {$L: "en", $d: Thu May 13 2021 23:59:59 GMT+0530 (India Standard Time), $x: {…}, $y: 2021, $M: 4, …}
```

When this situation arises the offset is 0 minutes, which is causing an infinite loop, hence freezing the website.

**Comment by baileypumfleet** [comment]
Unfortunately even on the latest deployment, it's still freezing

**Comment by parasg1999** [comment]
If you have a test DB, can I have access to that? It would be so much easier to debug

**Comment by baileypumfleet** [comment]
We don't have a publicly accessible test DB, but this is my 'user' row in the DB if any of that helps:
Username: bailey
Email: bailey@calendso.com
Timezone: GMT
Start Time: 480
End Time: 1380

**Comment by emrysal** [comment]
Hey @parasg1999 

I understand it's been a bit of time, I've merged main with your branch and attempted to PR to your branch; but for some reason Github doesn't find your branch.

Could you have a look at https://github.com/emrysal/calendso/tree/feature/slot-optimization ? I think the freezing problem is fixed in that branch (or at least, I don't have the issue).

File: lib/slots.ts [hunk 1]
```diff
@@ -12,17 +12,69 @@ const getMinutesFromMidnight = (date) => {
...
+    // Check if startTime is between slot 
+    else if (busyTimeStart.isBetween(slotStartTime, slotEndTime)) {
+      slotAvailability.available = false;
+      slotAvailability.offset = busyTimeEnd.diff(slotStartTime, 'minute');
+    }
+
+    if (!slotAvailability.available) {
+      if(slotAvailability.offset === 0) slotAvailability.offset = 5;
+      return slotAvailability
+    };
```
Comment (baileypumfleet) [inline lib/slots.ts:61]: Codacy has a fix for the issue: [Unnecessary semicolon.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
    }
```

File: lib/slots.ts [hunk 2]
```diff
@@ -12,17 +12,69 @@ const getMinutesFromMidnight = (date) => {
...
+    available: true,
+    offset: 0,
+  }
+
+  for (let i = 0; i < busyTimes.length; i += 1) {
+    const busyTimeStart = dayjs(busyTimes[i].start);
+    const busyTimeEnd = dayjs(busyTimes[i].end);
+
+    // Check if start times are the same
+    if (slotStartTime.format('HH:mm') == busyTimeStart.format('HH:mm')) {
```
Comment (baileypumfleet) [inline lib/slots.ts:35]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
    if (slotStartTime.format("HH:mm") == busyTimeStart.format('HH:mm')) {
```

File: lib/slots.ts [hunk 3]
```diff
@@ -12,17 +12,69 @@ const getMinutesFromMidnight = (date) => {
...
+    }
+
+    // Check if startTime is between slot 
+    else if (busyTimeStart.isBetween(slotStartTime, slotEndTime)) {
+      slotAvailability.available = false;
+      slotAvailability.offset = busyTimeEnd.diff(slotStartTime, 'minute');
+    }
+
+    if (!slotAvailability.available) {
+      if(slotAvailability.offset === 0) slotAvailability.offset = 5;
```
Comment (baileypumfleet) [inline lib/slots.ts:59]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
      if(slotAvailability.offset === 0) {slotAvailability.offset = 5;}
```

File: lib/slots.ts [hunk 4]
```diff
@@ -32,12 +84,19 @@ const getSlots = ({
     for (
       let minutes = dayStartTime;
       minutes <= dayEndTime - eventLength;
-      minutes += parseInt(eventLength, 10)
     ) {
       const slot = lowerBound.add(minutes, "minutes");
+
       if (slot > now) {
-        slots.push(slot);
+        const slotAvailablilty = isSlotAvailable({ slotStartTime: slot, slotEndTime: slot.add(eventLength, 'minutes'), busyTimes });
```
Comment (baileypumfleet) [inline lib/slots.ts:91]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
        const slotAvailablilty = isSlotAvailable({ slotStartTime: slot, slotEndTime: slot.add(eventLength, "minutes"), busyTimes });
```

File: lib/slots.ts [hunk 5]
```diff
@@ -12,17 +12,69 @@ const getMinutesFromMidnight = (date) => {
...
+  }
+
+  for (let i = 0; i < busyTimes.length; i += 1) {
+    const busyTimeStart = dayjs(busyTimes[i].start);
+    const busyTimeEnd = dayjs(busyTimes[i].end);
+
+    // Check if start times are the same
+    if (slotStartTime.format('HH:mm') == busyTimeStart.format('HH:mm')) {
+      slotAvailability.available = false;
+      slotAvailability.offset = busyTimeEnd.diff(slotStartTime, 'minute');
```
Comment (baileypumfleet) [inline lib/slots.ts:37]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
      slotAvailability.offset = busyTimeEnd.diff(slotStartTime, "minute");
```

File: lib/slots.ts (hunk 1 above)
Comment (baileypumfleet) [inline lib/slots.ts:61]: Codacy has a fix for the issue: [Unnecessary semicolon.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
    }
```

File: lib/slots.ts (hunk 2 above)
Comment (baileypumfleet) [inline lib/slots.ts:35]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
    if (slotStartTime.format("HH:mm") == busyTimeStart.format('HH:mm')) {
```

File: lib/slots.ts [hunk 6]
```diff
@@ -12,17 +12,69 @@ const getMinutesFromMidnight = (date) => {
...
   selectedDate,
   dayStartTime,
-  dayEndTime
+  dayEndTime,
+  busyTimes,
 }) => {
 
-  if(!selectedDate) return []
-  
+  if (!selectedDate) return []
```
Comment (baileypumfleet) [inline lib/slots.ts:76]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
  if(!selectedDate) {return []}
```

File: lib/slots.ts (hunk 3 above)
Comment (baileypumfleet) [inline lib/slots.ts:59]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
      if(slotAvailability.offset === 0) {slotAvailability.offset = 5;}
```

File: lib/slots.ts (hunk 4 above)
Comment (baileypumfleet) [inline lib/slots.ts:91]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
        const slotAvailablilty = isSlotAvailable({ slotStartTime: slot, slotEndTime: slot.add(eventLength, "minutes"), busyTimes });
```

File: lib/slots.ts (hunk 5 above)
Comment (baileypumfleet) [inline lib/slots.ts:37]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/calendso/calendso/pullRequest?prid=7689309)
```suggestion
      slotAvailability.offset = busyTimeEnd.diff(slotStartTime, "minute");
```

**Comment by baileypumfleet** [comment]
Is this made redundant by your more recent PR, @emrysal ?

**Pull request closed without merging**